src/render/painter.js

```diff
@@ -261,6 +274,87 @@ class Painter {
 
         this.frameHistory.record(Date.now(), this.transform.zoom, style.getTransition().duration);
 
+        const layerIds = this.style._order;
+
+        // 3D pass
+        // We first bind an offscreen framebuffer and render each 3D layer to
+        // its own texture, which we'll use later in the translucent pass to
+        // render to the main framebuffer. By doing this before we render to
+        // the main framebuffer we won't have to do an expensive framebuffer
+        // restore mid-render pass.
+        // The most important distinction of the 3D pass is that we use the
+        // depth buffer in an entirely different way (to represent 3D space)
+        // than we do in the 2D pass (to preserve layer order).
+        this.renderPass = '3d';
+        {
+            const gl = this.gl;
+            // We'll wait and only attach the framebuffer if we think we're
+            // rendering something.
+            let fboAttached = false;
+
+            let sourceCache;
+            let coords = [];
+
+            for (let i = 0; i < layerIds.length; i++) {
```

Comment:
Are there ways we could simplify the control flow in the body of this loop? For example, move checks such as `layer.paint['fill-extrusion-opacity'] === 0` into `has3DPass` and set `this._prerenderedTextures[layer.id] = null` up front for every layer? Or even eliminate `has3Dpass`, have the layers without a 3D pass simply do nothing, and have the 3D layers do the texture setup inside their `render` functions? Not sure what exactly is kosher performance-wise, but this code seems like it could be simpler.

Reply:
> have the 3D layers do the texture setup inside their `render` functions

Conceptually I don't prefer this — as it is, the `render` functions are mostly just responsible for rendering whatever they have to render to whatever target is bound, and the painter is responsible for managing render targets and switching between them. Especially if we add more types with `3d` passes I'd prefer to have the render target management for all 3D types happen in the painter's 3D pass loop. I did some simplification in https://github.com/mapbox/mapbox-gl-js/pull/5101/commits/677c2610a27c73df45f897a666bc3a69be082669.